File: backend/app/services/compactness.py
"""Strict Caveman compactness checks for prompt and memory authoring."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from app.services.compactness_policy import get_policy

ContentKind = Literal["prompt", "memory"]

FILLER_PATTERNS: tuple[tuple[str, re.Pattern[str]], ...] = (
    ("just", re.compile(r"\bjust\b", re.IGNORECASE)),
    ("really", re.compile(r"\breally\b", re.IGNORECASE)),
    ("basically", re.compile(r"\bbasically\b", re.IGNORECASE)),
    ("please", re.compile(r"\bplease\b", re.IGNORECASE)),
    ("let me know", re.compile(r"\blet me know\b", re.IGNORECASE)),
    ("feel free", re.compile(r"\bfeel free\b", re.IGNORECASE)),
    ("i recommend", re.compile(r"\bi recommend\b", re.IGNORECASE)),
    ("i suggest", re.compile(r"\bi suggest\b", re.IGNORECASE)),
    ("you should", re.compile(r"\byou should\b", re.IGNORECASE)),
    ("make sure", re.compile(r"\bmake sure\b", re.IGNORECASE)),
)
EXAMPLE_PATTERN = re.compile(r"\bfor example\b|\be\.g\.\b|example:", re.IGNORECASE)
HEDGE_PATTERN = re.compile(
    r"\b(?:maybe|probably|likely|might|could|should|usually|generally|try to)\b",
    re.IGNORECASE,
)
SOFT_TONE_PATTERN = re.compile(
    r"\b(?:be thorough|be objective|be specific|be precise|be natural|be conversational|be helpful|be friendly|be confident)\b",
    re.IGNORECASE,
)
OFFER_BACK_PATTERN = re.compile(
    r"\b(?:if you want|would you like|happy to help|happy to|let me know if)\b",
    re.IGNORECASE,
)
PROSE_CODE_BLOCK_PATTERN = re.compile(r"```.*?```", re.DOTALL)
INLINE_CODE_PATTERN = re.compile(r"`[^`]*`")
PLACEHOLDER_PATTERN = re.compile(r"\{[^{}\n]+\}")
WORD_PATTERN = re.compile(r"[A-Za-z']+")
ARTICLE_WORDS = {"a", "an", "the"}


@dataclass(frozen=True)
class CompactnessReport:
    kind: ContentKind
    chars: int
    lines: int
    tokens: int
    errors: tuple[str, ...]
    warnings: tuple[str, ...]


class CompactnessValidationError(Exception):
    """Raised when content violates the strict Caveman gate."""

    def __init__(self, kind: ContentKind, errors: list[str]):
        self.kind = kind
        self.errors = tuple(errors)
        detail = "; ".join(errors)
        super().__init__(f"{kind} failed strict Caveman gate: {detail}")


def _line_count(content: str) -> int:
    return content.count("\n") + (1 if content and not content.endswith("\n") else 0)


def _estimate_tokens(content: str) -> int:
    return 0 if not content else max(1, len(content) // 4)


def _detect_fillers(content: str) -> list[str]:
    return [label for label, pattern in FILLER_PATTERNS if pattern.search(content)]


def _strip_non_prose(content: str) -> str:
    stripped = PROSE_CODE_BLOCK_PATTERN.sub(" ", content)
    stripped = INLINE_CODE_PATTERN.sub(" ", stripped)
    stripped = PLACEHOLDER_PATTERN.sub(" ", stripped)
    stripped = re.sub(r"(?m)^\s{0,3}#+\s*", "", stripped)
    stripped = re.sub(r"(?m)^\s*[-*]\s+", "", stripped)
    stripped = re.sub(r"(?m)^\s*\d+\.\s+", "", stripped)
    return stripped


def _extract_sentences(content: str) -> list[str]:
    prose = _strip_non_prose(content)
    return [
        sentence.strip()
        for sentence in re.split(r"(?<=[.!?])\s+|\n+", prose)
        if WORD_PATTERN.search(sentence)
    ]


def _article_ratio(words: list[str]) -> float:
    if not words:
        return 0.0
    article_count = sum(1 for word in words if word in ARTICLE_WORDS)
    return article_count / len(words)


def analyze_compactness(content: str, *, kind: ContentKind) -> CompactnessReport:
    policy = get_policy()
    chars = len(content)
    lines = _line_count(content)
    tokens = _estimate_tokens(content)
    errors: list[str] = []
    warnings: list[str] = []
    filler_hits = _detect_fillers(content)
    sentences = _extract_sentences(content)
    prose_words = WORD_PATTERN.findall(_strip_non_prose(content).lower())
    article_ratio = _article_ratio(prose_words)

    if kind == "prompt":
        if tokens > policy.prompt_max_tokens:
            warnings.append(f"large prompt ({tokens} tok). Hot-path prompts pay this every turn.")
        if lines > policy.prompt_max_lines:
            warnings.append(f"long prompt ({lines} lines). Collapse repeated examples and overlap.")
    else:
        if chars > policy.memory_max_chars:
            warnings.append(f"long memory ({chars} chars). Keep one atomic rule; split if needed.")
        if lines > policy.memory_max_lines:
            warnings.append(f"multi-line memory ({lines} lines). Prefer one short rule body.")

    if filler_hits:
        errors.append(f"filler terms found: {', '.join(filler_hits[:4])}")

    if EXAMPLE_PATTERN.search(content):
        errors.append("example markers found. Strip examples; keep direct rules only.")

    if HEDGE_PATTERN.search(content):
        errors.append("hedging found. Replace maybe/should/could-style phrasing with direct rules.")

    if SOFT_TONE_PATTERN.search(content):
        errors.append("soft-tone phrasing found. Replace 'be X' guidance with direct action rules.")

    if OFFER_BACK_PATTERN.search(content):
        errors.append("offer-back phrasing found. Remove optional follow-up or helper language.")

    if (
        prose_words
        and len(prose_words) >= policy.article_ratio_min_words
        and article_ratio > policy.max_article_ratio
    ):
        errors.append(
            f"article-heavy prose ({article_ratio:.1%}). Drop articles and compress sentence shape."
        )

    long_sentences = [
        sentence
        for sentence in sentences
        if len(WORD_PATTERN.findall(sentence)) > policy.max_sentence_words
    ]
    if long_sentences:
        errors.append("long prose sentences found. Split into short direct lines or bullets.")

    if sentences:
        average_sentence_words = sum(
            len(WORD_PATTERN.findall(sentence)) for sentence in sentences
        ) / len(sentences)
        if (
            len(prose_words) >= policy.avg_sentence_min_words
            and average_sentence_words > policy.max_avg_sentence_words
        ):
            errors.append(
                f"average sentence too long ({average_sentence_words:.1f} words). Compress prose."
            )

    return CompactnessReport(
        kind=kind,
        chars=chars,
        lines=lines,
        tokens=tokens,
        errors=tuple(errors),
        warnings=tuple(warnings),
    )


def validate_compactness(content: str, *, kind: ContentKind) -> CompactnessReport:
    report = analyze_compactness(content, kind=kind)
    if report.errors:
        raise CompactnessValidationError(kind, list(report.errors))
    return report

File: frontend/src/lib/compactness.ts
export type CompactnessKind = 'prompt' | 'memory'

export interface CompactnessReport {
  kind: CompactnessKind
  chars: number
  lines: number
  tokens: number
  errors: string[]
  warnings: string[]
}

export interface CompactnessThresholds {
  memory_max_chars: number
  memory_max_lines: number
  prompt_max_tokens: number
  prompt_max_lines: number
  max_sentence_words: number
  max_avg_sentence_words: number
  avg_sentence_min_words: number
  max_article_ratio_permille: number
  article_ratio_min_words: number
}

export const COMPACTNESS_DEFAULTS: CompactnessThresholds = {
  memory_max_chars: 280,
  memory_max_lines: 4,
  prompt_max_tokens: 350,
  prompt_max_lines: 80,
  max_sentence_words: 24,
  max_avg_sentence_words: 16,
  avg_sentence_min_words: 120,
  max_article_ratio_permille: 85,
  article_ratio_min_words: 80,
}

const FILLER_PATTERNS: Array<[string, RegExp]> = [
  ['just', /\bjust\b/i],
  ['really', /\breally\b/i],
  ['basically', /\bbasically\b/i],
  ['please', /\bplease\b/i],
  ['let me know', /\blet me know\b/i],
  ['feel free', /\bfeel free\b/i],
  ['i recommend', /\bi recommend\b/i],
  ['i suggest', /\bi suggest\b/i],
  ['you should', /\byou should\b/i],
  ['make sure', /\bmake sure\b/i],
]
const EXAMPLE_PATTERN = /\bfor example\b|\be\.g\.\b|example:/gi
const HEDGE_PATTERN =
  /\b(?:maybe|probably|likely|might|could|should|usually|generally|try to)\b/i
const SOFT_TONE_PATTERN =
  /\b(?:be thorough|be objective|be specific|be precise|be natural|be conversational|be helpful|be friendly|be confident)\b/i
const OFFER_BACK_PATTERN =
  /\b(?:if you want|would you like|happy to help|happy to|let me know if)\b/i
const PROSE_CODE_BLOCK_PATTERN = /```[\s\S]*?```/g
const INLINE_CODE_PATTERN = /`[^`]*`/g
const PLACEHOLDER_PATTERN = /\{[^{}\n]+\}/g
const WORD_PATTERN = /[A-Za-z']+/g
const ARTICLE_WORDS = new Set(['a', 'an', 'the'])

function lineCount(content: string): number {
  return content ? content.split('\n').length : 0
}

function estimateTokens(content: string): number {
  return content ? Math.ceil(content.length / 4) : 0
}

function detectFillers(content: string): string[] {
  return FILLER_PATTERNS.filter(([, pattern]) => pattern.test(content)).map(
    ([label]) => label,
  )
}

function stripNonProse(content: string): string {
  return content
    .replace(PROSE_CODE_BLOCK_PATTERN, ' ')
    .replace(INLINE_CODE_PATTERN, ' ')
    .replace(PLACEHOLDER_PATTERN, ' ')
    .replace(/^\s{0,3}#+\s*/gm, '')
    .replace(/^\s*[-*]\s+/gm, '')
    .replace(/^\s*\d+\.\s+/gm, '')
}

function proseWords(content: string): string[] {
  return stripNonProse(content).toLowerCase().match(WORD_PATTERN) ?? []
}

function extractSentences(content: string): string[] {
  return stripNonProse(content)
    .split(/(?<=[.!?])\s+|\n+/)
    .map((value) => value.trim())
    .filter((value) => WORD_PATTERN.test(value))
}

export function analyzeCompactness(
  content: string,
  kind: CompactnessKind,
  policy: CompactnessThresholds = COMPACTNESS_DEFAULTS,
): CompactnessReport {
  const chars = content.length
  const lines = lineCount(content)
  const tokens = estimateTokens(content)
  const errors: string[] = []
  const warnings: string[] = []
  const fillerHits = detectFillers(content)
  const words = proseWords(content)
  const sentences = extractSentences(content)

  if (kind === 'prompt') {
    if (tokens > policy.prompt_max_tokens) {
      warnings.push(
        `large prompt (${tokens} tok). Hot-path prompts pay this every turn.`,
      )
    }
    if (lines > policy.prompt_max_lines) {
      warnings.push(
        `long prompt (${lines} lines). Collapse repeated examples and overlap.`,
      )
    }
  } else {
    if (chars > policy.memory_max_chars) {
      warnings.push(
        `long memory (${chars} chars). Keep one atomic rule; split if needed.`,
      )
    }
    if (lines > policy.memory_max_lines) {
      warnings.push(
        `multi-line memory (${lines} lines). Prefer one short rule body.`,
      )
    }
  }

  if (fillerHits.length > 0) {
    errors.push(`filler terms found: ${fillerHits.slice(0, 4).join(', ')}`)
  }

  if (EXAMPLE_PATTERN.test(content)) {
    errors.push(
      'example markers found. Strip examples; keep direct rules only.',
    )
  }

  if (HEDGE_PATTERN.test(content)) {
    errors.push(
      'hedging found. Replace maybe/should/could-style phrasing with direct rules.',
    )
  }

  if (SOFT_TONE_PATTERN.test(content)) {
    errors.push(
      "soft-tone phrasing found. Replace 'be X' guidance with direct action rules.",
    )
  }

  if (OFFER_BACK_PATTERN.test(content)) {
    errors.push(
      'offer-back phrasing found. Remove optional follow-up or helper language.',
    )
  }

  if (words.length >= policy.article_ratio_min_words) {
    const articleRatio =
      words.filter((word) => ARTICLE_WORDS.has(word)).length / words.length
    if (articleRatio > policy.max_article_ratio_permille / 1000) {
      errors.push(
        `article-heavy prose (${(articleRatio * 100).toFixed(1)}%). Drop articles and compress sentence shape.`,
      )
    }
  }

  const sentenceLengths = sentences.map(
    (sentence) => sentence.match(WORD_PATTERN)?.length ?? 0,
  )
  if (sentenceLengths.some((count) => count > policy.max_sentence_words)) {
    errors.push(
      'long prose sentences found. Split into short direct lines or bullets.',
    )
  }

  if (
    words.length >= policy.avg_sentence_min_words &&
    sentenceLengths.length > 0
  ) {
    const averageSentenceWords =
      sentenceLengths.reduce((sum, count) => sum + count, 0) /
      sentenceLengths.length
    if (averageSentenceWords > policy.max_avg_sentence_words) {
      errors.push(
        `average sentence too long (${averageSentenceWords.toFixed(1)} words). Compress prose.`,
      )
    }
  }

  return { kind, chars, lines, tokens, errors, warnings }
}
